proposals/proposals

```diff
@@ -0,0 +1,214 @@
+> - **The resulting set of candidate methods is grouped by ***overload_resolution_priority***. All candidates not in the highest (by standard integer comparison) ***overload_resolution_priority*** group are removed from the set.**
+> - The best method of the set of candidate methods is identified using the overload resolution rules of [§12.6.4](expressions.md#1264-overload-resolution). If a single best method cannot be identified, the method invocation is ambiguous, and a binding-time error occurs. When performing overload resolution, the parameters of a generic method are considered after substituting the type arguments (supplied or inferred) for the corresponding method type parameters.
+
+As an example, this feature would cause the following code snippet to print "Span", rather than "Array":
+
+```cs
+using System.Runtime.CompilerServices;
+
+var d = new C1();
+int[] arr = [1, 2, 3];
+d.M(arr); // Prints "Span"
+
+class C1
+{
+    [OverloadResolutionPriority(1)]
+    public void M(ReadOnlySpan<int> s) => Console.WriteLine("Span");
+    // Default overload resolution priority
+    public void M(int[] a) => Console.WriteLine("Array");
+}
+```
+
+The effect of this change is that, like pruning for most-derived types, we add a final pruning for overload resolution priority. Because this pruning occurs at the very end of the overload resolution
+process, it does mean that a base type cannot make its members higher-priority than any derived type. This is intentional, and prevents an arms-race from occuring where a base type may try to always
+be better than a derived type. For example:
+
+```cs
+using System.Runtime.CompilerServices;
+
+var d = new Derived();
+d.M([1, 2, 3]); // Prints "Derived", because members from Base are not considered due to finding an applicable member in Derived
+
+class Base
+{
+    [OverloadResolutionPriority(1)]
+    public void M(ReadOnlySpan<int> s) => Console.WriteLine("Base");
+}
+
+class Derived : Base
+{
+    public void M(int[] a) => Console.WriteLine("Derived");
+}
+```
+
+Negative numbers are allowed to be used, and can be used to mark a specific overload as worse than all other default overloads.
+
+### `System.Runtime.CompilerServices.OverloadResolutionPriorityAttribute`
+
+We introduce the following attribute to the BCL:
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+[AttributeUsage(AttributeTargets.Method AllowMultiple = false, Inherited = false)]
+public sealed class OverloadResolutionPriorityAttribute(int priority)
+{
+    public int Priority => priority;
+}
+```
+
+**Open question**: Should the attribute be inherited? If not, what is the priority of the overriding member?  
+**Open question**: If the attribute is specified on a virtual member, should an override of that member be required to repeat the attribute?  
+**Open question**: Should the attribute support constructors?
```

Comment:
My opinions are:
- For inheritance, it should just default to 0. It shouldn't warn if one isn't specified, but it would be fine to have a suggestion. Rationale: The specific value is only relevant within a specific class, so it doesn't make much sense to me to require, or even warn, on missing value, since the chosen values are only relevant within a specific class.
- For constructors: it would be good if it supported constructors, and indexers also. It should potentially likewise be disallowed on property getters & setters, event `add` and `remove` also.

Reply:
> since all derived overloads are considered first

This is actually not how overload resolution works.

Reply:
I was referring to:

> The set of candidate methods is reduced to contain only methods from the most derived types

Have I misinterpreted it? If so, what does it mean, so I can understand how it works?

I interpreted it as: once you have a list of methods you could call, only consider the definitions in the most derived type that defines at least one of them (i.e., don't consider a virtual override in this list, since it's not a 'definition').

Reply:
> I interpreted it as: once you have a list of methods you could call, only consider the definitions in the most derived type that defines at least one of them (i.e., don't consider a virtual override in this list, since it's not a 'definition').

You're almost there. The key thing is that overrides are _not_ definitions in the most derived type. You can see that in [this](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBQG4TIAEGdAvHTAKYDudAIm8lFTYYAFAEoA3BVgI6UZnWKTyGAHQBZYVAl0A9DroAFfjAQBnOgCIANgHsYAcwsUK+AMx0IwUwmQQAxjJEdABCEKZsdADeFHSxdG4eXj7+gQAsdBrSshIUAL7O5AlBvPyCjCAhYRHR5HHx7jaCyPwYEfjpmSbZMXE1dXX4xACcwhbSFjm1cflTsT2xRR3Ctg50VqLzUZsDw6MrjpN1M7lAA==) example:

```cs
using System;

var d = new Derived();
int i = 1;
d.M(i); // Prints "long"

public abstract class Base {
    public abstract void M(int i);
}

public class Derived : Base {
    public override void M(int i)
    {
        Console.WriteLine("int");
    }
    
    public void M(long l)
    {
        Console.WriteLine("long");
    }
}
```